patches/tModLoader/Terraria/ID/SetFactory.TML.cs

```diff
@@ -0,0 +1,159 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using ReLogic.Utilities;
+using Terraria.ModLoader;
+using Terraria.ModLoader.Core;
+
+namespace Terraria.ID;
+
+public partial class SetFactory
+{
+	// Additional code to support named custom sets for ad-hoc collaboration
+	private record SetMetadata(Type type, object defaultValue); // Note: Stores the array Type, not Type[]
+
+	private ConcurrentDictionary<(string, Type), SetMetadata> setMetadataMapping = new ConcurrentDictionary<(string, Type), SetMetadata>();
+
+	public void Clear() // TODO: call where?
+	{
+		setMetadataMapping.Clear();
+	}
+
+	// Copies of existing methods with an additional key parameter.
+	public T[] CreateNamedCustomSet<T>(string key, T defaultState, params object[] inputs)
+	{
+		var set = CreateCustomSet(defaultState, inputs);
+		RegisterNamedCustomSet(key, defaultState, ref set);
+		return set;
+	}
+
+	public float[] CreateNamedFloatSet(string key, float defaultState, params float[] inputs)
+	{
+		var set = CreateFloatSet(defaultState, inputs);
+		RegisterNamedCustomSet(key, defaultState, ref set);
+		return set;
+	}
+	public ushort[] CreateNamedUshortSet(string key, ushort defaultState, params ushort[] inputs)
+	{
+		var set = CreateUshortSet(defaultState, inputs);
+		RegisterNamedCustomSet(key, defaultState, ref set);
+		return set;
+	}
+	public int[] CreateNamedIntSet(string key, int defaultState, params int[] inputs)
+	{
+		var set = CreateIntSet(defaultState, inputs);
+		RegisterNamedCustomSet(key, defaultState, ref set);
+		return set;
+	}
+	public int[] CreateNamedIntSet(string key, params int[] types)
+	{
+		var set = CreateIntSet(types);
+		RegisterNamedCustomSet(key, -1, ref set);
+		return set;
```

Comment:
You could make this single line by replacing the `ref` with a regular parameter, and having `RegisterNamedCustomSet` return the merged result